testnet/stacks-node/src/tests/signer.rs

```diff
@@ -0,0 +1,249 @@
+use stacks::chainstate::stacks::StacksPrivateKey;
+use stacks_signer::runloop::RunLoopCommand;
+
+const SLOTS_PER_USER: u32 = 16;
+
+fn build_contract(num_signers: u32, signer_stacks_private_keys: &[StacksPrivateKey]) -> String {
+    let mut stackerdb_contract = String::new(); // "
+    stackerdb_contract += "        ;; stacker DB\n";
+    stackerdb_contract += "        (define-read-only (stackerdb-get-signer-slots)\n";
+    stackerdb_contract += "            (ok (list\n";
+    for i in 0..num_signers {
+        stackerdb_contract += "                {\n";
+        stackerdb_contract += format!(
+            "                    signer: '{},\n",
+            to_addr(&signer_stacks_private_keys[i as usize])
+        )
+        .as_str();
+        stackerdb_contract +=
+            format!("                    num-slots: u{}\n", SLOTS_PER_USER).as_str();
+        stackerdb_contract += "                }\n";
+    }
+    stackerdb_contract += "                )))\n";
+    stackerdb_contract += "\n";
+    stackerdb_contract += "        (define-read-only (stackerdb-get-config)\n";
+    stackerdb_contract += "            (ok {\n";
+    stackerdb_contract += "                chunk-size: u4096,\n";
+    stackerdb_contract += "                write-freq: u0,\n";
+    stackerdb_contract += "                max-writes: u4096,\n";
+    stackerdb_contract += "                max-neighbors: u32,\n";
+    stackerdb_contract += "                hint-replicas: (list )\n";
+    stackerdb_contract += "            }))\n";
+    stackerdb_contract += "    ";
+
+    info!("stackerdb_contract:\n{}\n", &stackerdb_contract);
+    stackerdb_contract
+}
+
+fn build_signer_config_tomls(
+    num_signers: u32,
+    signer_stacks_private_keys: &[StacksPrivateKey],
+    contract_id: String,
+) -> Vec<String> {
+    let mut rng = OsRng::default();
+    let num_keys: u32 = 20;
+    let keys_per_signer = num_keys / num_signers;
+    let mut key_id: u32 = 1;
+    let mut key_ids = Vec::new();
+    for _ in 0..num_signers {
+        let mut ids = Vec::new();
+        for _ in 0..keys_per_signer {
+            ids.push(format!("{key_id}"));
+            key_id += 1;
+        }
+        key_ids.push(ids.join(", "));
+    }
+    let signer_ecdsa_private_keys = (0..num_signers)
+        .map(|_| Scalar::random(&mut rng))
+        .collect::<Vec<Scalar>>();
+
+    let mut signer_config_tomls = vec![];
+    let mut signers_array = String::new();
+    signers_array += "signers = [";
+    for (i, private_key) in signer_ecdsa_private_keys.iter().enumerate() {
+        let ecdsa_public_key = ecdsa::PublicKey::new(private_key).unwrap().to_string();
+        let ids = key_ids[i].clone();
+        signers_array += &format!(
+            r#"
+            {{public_key = "{ecdsa_public_key}", key_ids = [{ids}]}}
+        "#
+        );
+        if i != signer_ecdsa_private_keys.len() - 1 {
+            signers_array += ",";
+        }
+    }
+    signers_array += "]";
+    let mut port = 30000;
```

Comment:
I think this needs to be the port from `config.node.rpc_bind`, since `node_host` gets passed to `libstackerdb` the RPC endpoint for the node.

Reply:
ah so this means I need to spin up num_signers neon nodes. They cannot just use the same one as we will have port conflicts if I try to bind multiple times to the same port.

Reply:
If you look at the current code, I'm using the rpc_bind from stacks node config as node_host, and then have a separate endpoint config key that uses the 3000x ports for the stackerdb receivers